True Layer Load Cancel

james-rae opened the pull request:
## Related Item(s)

donethankses #1491 #2199 #2200

## Changes

`tldr` Implements true layer cancellation. Removes lots of timeouts and visibility trickery that managed ESRI layers doing as they pleased. Layer now drives cancellation instead of the legend, simplifying more code.

- Added layer method `cancelLoad()`
  - Added cancel kickouts to every asynch step of layer initialize / load process
  - ESRI layers no longer wait in secret to finish loaded, then cancel
- Added safety kickouts to legend cleanups to avoid forever-spinning intervals.
- Changes to timeout limits on layer loading now that we have instant cancel power.
  - Merged the layer `maxLoadTime` timeout and hardcoded 20 seconds initiation timer on `MapAPI.addLayer()`.
  - Boosted map default from 20 seconds to 90 seconds.
  - Changed layer default to inherit from the map default.
    - This makes it easier to globally adjust a load threshold. Otherwise every layer needs to be changed, and wizard layers will disrespect.
  - Both timeout values are now configurable (per instance, per layer, combos).
- Feature, File, and WFS now report their `layerIdx` instead of `-1`.
- Emit Layer Registered event when it registers (was delaying until it loaded).
- Emit Reload Start event regardless of current layer state (was only doing it for initialized layers).
- Layer Load Status Change event now has additional property on the payload so a listener can determine if an Error report was due to a user cancelling a layer.
- Fixed bad falsey logic in legend surrounding MIL sublayer indexes being 0.
- Fixed some other legend bugs that I didn't log.
- Removed a lot of duplicate method doc.
- Corrected an alphabet crime.


## Notes

The delay on layer initiation (and other testing trickery) will be removed prior to merging. Everything is marked with `~@~` for easy finding.

If cancelling an ESRI server layer during the faked delay of the Load phase, ESRI will screech errors to the console. This is out of our control, but is also a manufactured scenario. In most real cases where layer needs to be cancelled it's stalled, so errors are legit. Also only `F12` devs will notice. 

### Breaking Changes to put in Release Notes

- The `layer/registered` event is now raised at registration, not post-initiation. Previously these two things happened at the same time. The doc page unfortunately used the words "added to map" which is the big lie and what makes this breaking.
- The `layer/reloadstart` event is now always raised when a reload starts. Previously it would only fire if the layer had initiated. Since a layer can be cancelled or fail prior to that, this didn't make sense.
- With no configuration, the maximum load time before erroring is now 90 seconds. It was previously 20 seconds. This number is configurable at both layer and map level.
- The layer config property `maxLoadTime` no longer supports infinite load (`0`). Setting an enormous value can approximate. This can be done at the layer or map configuration level.

## QA Testing

Please use the updated QA PR
https://github.com/ramp4-pcar4/ramp4-pcar4/pull/2391

## Testing

This change covers a lot of ground, so no "simple steps". Will present a bunch of ways to try to break things.

### General

All samples will ping notifications when the init and load phases finish for a layer, so you can watch for the number icon to bump if you want to ensure you cancel during a specific phase.

Enhanced Sample 3, 4, 5, and 7 have delays on both initialize and load phases set to 4 seconds. This will apply to anything added via the wizard as well.

These four samples will help cover the main scenarios.

- Sample 3: Feature layer in config
- Sample 4: MIL children in config
- Sample 5: File layer in config
- Sample 7: Layer medley, including WMS. Also good for testing reorder fixture with cancelled layers.

Reminder that Feature & MIL layers have very simple initialization, and more work in the load phase. File & WFS do heavy lifting in initialization, and have basic load work. But it's important to try to break all combos.

Other layers can be added via the wizard for more tests. If anyone wants custom delays on their wizard-added layers, they can mod stuff in the console.

```
debugInstance.loadDelay = 8000;
debugInstance.initDelay = 7000;
```

### Load Timeout

To test the load timeout hard-limit, dropping this in the console of Enhanced Sample 3 will simulate it.

```
const failfastConf = {
	id: 'iFailFast',
	name: 'I will always fail',
	layerType: 'ogc-wfs',
	url: 'https://api.weather.gc.ca//collections/ahccd-trends/items?measurement_type__type_mesure=total_precip&period__periode=Ann&offset=0&limit=150&province__province=on',
	maxLoadTime: 400
};

const failFastLyr = debugInstance.geo.layer.createLayer(failfastConf);
debugInstance.geo.map.addLayer(failFastLyr); // will put uncaught promise error on console. all good
debugInstance.event.emit('user/layeradded', failFastLyr); // adds to legend
```

### MIL Child Detection

A tricky area is legend configs waiting for MIL children who don't exist until after the parent loads.

Enhanced Sample 4 is the best to test this on. Both legend entries are bound to children that won't exist until various phases complete.

### MIL TreeGrow

To test the growing of MIL tree structures, run this code in the console for Enhanced Sample 4.

```
const treeGrowConf = {
	id: 'iGrowTrees',
	name: 'Tree Grower Layer',
	layerType: 'esri-map-image',
	url: 'https://section917.canadacentral.cloudapp.azure.com/arcgis/rest/services/TestData/Nest/MapServer',
	sublayers: [{ index: 0 }]
	
};

const treeGrowLyr = debugInstance.geo.layer.createLayer(treeGrowConf);
debugInstance.geo.map.addLayer(treeGrowLyr); // may put uncaught promise error on console. all good
debugInstance.event.emit('user/layeradded', treeGrowLyr); // adds to legend
```

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)
<!-- Reviewable:end -->

james-rae commented:
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-O3iJIcs3E2tiNFw-YK0):*
Blocker. Remove everything with `~@~`


<!-- Sent from Reviewable.io -->

spencerwahl reviewed (changes requested):
> Reviewed 14 of 45 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: 14 of 45 files reviewed, 2 unresolved discussions (waiting on @james-rae)

___
*[`src/fixtures/legend/components/item.vue` line 819 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-O3ib_B48pykZQcqHE6P:-O3ib_B48pykZQcqHE6Q:bkf0nva) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/8a3a2c9459aa0e115eb9e4dbb035426d0122063a/src/fixtures/legend/components/item.vue#L819)):*
> ```Vue
> const cancelOrRemoveLayer = () => {
>     const layerItem: LayerItem = toRaw(props.legendItem as LayerItem); // so that typescript doesn't yell in the whole method
>     let safteyCount = 0;
> ```

`safetyCount` very important to fix


<!-- Sent from Reviewable.io -->

james-rae commented:
*[`src/fixtures/legend/components/item.vue` line 819 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-O3ib_B48pykZQcqHE6P:-O3m2f81DUwpcDSbTrtg:b14a5lg) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/8a3a2c9459aa0e115eb9e4dbb035426d0122063a/src/fixtures/legend/components/item.vue#L819)):*
<details><summary><i>Previously, spencerwahl (Spencer Wahl) wrote…</i></summary><blockquote>

`safetyCount` very important to fix
</blockquote></details>

good catch. this could have bricked most prod sites.


<!-- Sent from Reviewable.io -->

szczz reviewed (changes requested):
A bunch of console grousin occurs when you remove a loaded MIL parent. 

Steps to reproduce:
1. Open enhanced sample 4 
2. Open the console
3. Remove one of the default layers once its loaded and notice no grousin
4. Add this layer [https://section917.canadacentral.cloudapp.azure.com/arcgis/rest/services/TestData/EcoAction/MapServer](https://section917.canadacentral.cloudapp.azure.com/arcgis/rest/services/TestData/EcoAction/MapServer) 
5. Remove one of the child layers and notice no grousin
6. Remove the parent (Layers) and watch the console light up

> Reviewed 43 of 45 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-:-O3mXNqdC1UoqYqqvFaQ:bq3zbuw)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae and @spencerwahl)



<!-- Sent from Reviewable.io -->

james-rae commented:
Good testin'.
Cleaned up some MIL parent-child synch code that was being fast and loose.
Those console errors were put there at the beginning when we wanted to know about bad code. Given Vue bindings can do reads whenever they please, and layers can now easily flit between initialized and not, I've dropped the warnings from the harmless getters (they still return the same benign defaults as before). Setters and dangerous getters will still screech.





<!-- Sent from Reviewable.io -->

szczz reviewed (approved):
> Reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae and @spencerwahl)



<!-- Sent from Reviewable.io -->

gordlin reviewed (approved):
> Reviewed 36 of 45 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae and @spencerwahl)



<!-- Sent from Reviewable.io -->

IshavSohal reviewed (approved):
> Reviewed 36 of 45 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae and @spencerwahl)



<!-- Sent from Reviewable.io -->

mohsin-r reviewed (changes requested):
Nice! Really liked the time comparing idea to stop the load actions in their tracks.

One bug I found was that if you cancel the MILs on sample 4 before they finish initiation and then hit reload, they will reload forever. It works properly if you cancel and reload between initiation and load, however.

> Reviewed 36 of 45 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-:-O4695W34whXy-p37n7O:b-rkv29)* status: all files reviewed, 3 unresolved discussions (waiting on @james-rae and @spencerwahl)

___
*[`src/api/panel-instance.ts` line 15 at r4](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-O46A7GCD8aXtGPnsUKS:-O46A7GDD_rlfBrhwuMg:b-51cdkf) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/b1e5b924a21b3b4bc339ee3bb70640a3fc723d9e/src/api/panel-instance.ts#L15)):*
> ```XML
> import type {
>     AsyncComponentFactoryEh,
>     AsyncComponentEh,
> ```

The alphabet crime is still there.


<!-- Sent from Reviewable.io -->

spencerwahl reviewed (commented):
^ reproduced. Definitely something to do with cancelling a MIL sublayer as cancelling init on sample 3 works fine.

> Reviewed 23 of 45 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-:-O46hgvx1HUvMJ01HcFJ:b-ozpndl)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae)



<!-- Sent from Reviewable.io -->

james-rae reviewed (commented):
Good testing.
Root cause was lurking assumption about the existence of child layer instances. Cancelling during initiation meant they didn't exist yet, causing the boom.
- Debug exposed some other scenarios like that, so fixed as well.
- Previous fix for Matt's bug was also a tad too cute, corrected.
- Weird Vue disrespect also uncovered a very old type bug in quick cache that was being very inefficient for point geometry caching, now fixed.

> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-:-O4LOyiWC8nLm3gxryhO:b-3xpic)* status: 37 of 48 files reviewed, 2 unresolved discussions (waiting on @gordlin, @IshavSohal, @mohsin-r, @spencerwahl, and @szczz)

___
*[`src/api/panel-instance.ts` line 15 at r4](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-O46A7GCD8aXtGPnsUKS:-O46_mYN2mCyONPT1knM:bhq5ew4) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/b1e5b924a21b3b4bc339ee3bb70640a3fc723d9e/src/api/panel-instance.ts#L15)):*
<details><summary><i>Previously, mohsin-r (Mohsin Reza) wrote…</i></summary><blockquote>

The alphabet crime is still there.
</blockquote></details>

That is just terrible work on my part.


<!-- Sent from Reviewable.io -->

spencerwahl reviewed (approved):
> Reviewed 10 of 11 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae and @mohsin-r)



<!-- Sent from Reviewable.io -->

mohsin-r reviewed (changes requested):
> Reviewed 10 of 11 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae)

___
*[`src/fixtures/legend/api/legend.ts` line 335 at r6](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-O4MJ779AtPZtz6tCP3F:-O4MJ779AtPZtz6tCP3G:b2nvzz6) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/7c31ba6242dc65dc9f610f2fa3022bf17f34cbb2/src/fixtures/legend/api/legend.ts#L335)):*
> ```XML
>         // sad, was hoping TS 5.5 would be smart enough to know the filter is all LayerItem
>         // @ts-ignore
>         return fullLegend.filter(
> ```

If I am correctly understanding what this method is supposed to do, this needs to be a full tree search as opposed to just searching the top root level. Otherwise, if I have a single layer item tied to a layer at depth 2, it won't show up.


<!-- Sent from Reviewable.io -->

james-rae reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @mohsin-r)

___
*[`src/fixtures/legend/api/legend.ts` line 335 at r6](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-O4MJ779AtPZtz6tCP3F:-O4MWNVjAmhHHOJC0aHZ:b-gzst7x) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/7c31ba6242dc65dc9f610f2fa3022bf17f34cbb2/src/fixtures/legend/api/legend.ts#L335)):*
<details><summary><i>Previously, mohsin-r (Mohsin Reza) wrote…</i></summary><blockquote>

If I am correctly understanding what this method is supposed to do, this needs to be a full tree search as opposed to just searching the top root level. Otherwise, if I have a single layer item tied to a layer at depth 2, it won't show up.
</blockquote></details>

Excellent eyes (and bad method doc reading by me). Corrected, and refactored a bunch of stuff for fun & profit.


<!-- Sent from Reviewable.io -->

spencerwahl reviewed (approved):
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 2 unresolved discussions (waiting on @james-rae and @mohsin-r)



<!-- Sent from Reviewable.io -->

mohsin-r reviewed (approved):
Was doing a sanity check and came across a minor issue. On enhanced sample 4, if you cancel the sublayers and then remove only one of them while reloading the other one, both sublayers appear on the map but the correct one disappears after a second or so. Not blocking as it technically works but doesn't look ideal and could be a new issue.

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-:-O4QPB5XAJw1UrwCVI-k:b-be438u)* status: all files reviewed, 1 unresolved discussion (waiting on @james-rae)



<!-- Sent from Reviewable.io -->

james-rae commented:
Yeah I noticed that as well when I was trying to break things.
I suspect it's due to one of these things:
- the 400ms delay on the reload button to ensure some legend animation happens before an immediate re-fail on reload
- our fake delays in initiation mean the sublayers delay. Parent might be doing initial draw based on config + service, doesn't realize child is removed until delay ends. Sublayer init is typically instant, so will be curious to see if scenario goes away after I remove the `sleep()` calls.
- I _thought_ sublayer visibility changes had a debounce to prevent multiple re-draws if a group block was toggled, but I can't find it. Might still be there, and hiding.

Anyway, sure, new issue with low priority makes sense. That click sequence will be very rare for typical user. And hopefully it vanishes after fake delays are removed.





<!-- Sent from Reviewable.io -->

mohsin-r reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 1 unresolved discussion (waiting on @james-rae)



<!-- Sent from Reviewable.io -->

dnlnashed reviewed (approved):
Noticed that when adding an unreloadable file layer through the wizard, if you cancel it during either phase you have the option still to reload it, infinitely loading the layer. Not blocking since this can be reproduced outside of this PR, can just be a new issue.

> Reviewed 32 of 45 files at r1, 5 of 7 files at r4, 10 of 11 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314#-:-O4QlH_n1q9b-2FPddse:b-wpvln2)* status: all files reviewed, 1 unresolved discussion (waiting on @james-rae)



<!-- Sent from Reviewable.io -->

james-rae commented:
Good catch. This PR needs to marinate for a week at least so I can probably fix it. Looking at the original "no reload" PR it appears it only focused on the menu button, not the "I failed during load" reload button in the legend block. 
Will need to figure out what makes the most sense; can get some weird scenarios, but usually only around sneaky layers using `rawData`





<!-- Sent from Reviewable.io -->

james-rae commented:
Enhanced. Also made some of the tree searchers in `LegendAPI` public, as they're useful. And then did some refactors.





<!-- Sent from Reviewable.io -->

mohsin-r reviewed (approved):
> Reviewed 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 1 unresolved discussion (waiting on @james-rae)



<!-- Sent from Reviewable.io -->

dnlnashed reviewed (approved):
> Reviewed 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: all files reviewed, 1 unresolved discussion (waiting on @james-rae)



<!-- Sent from Reviewable.io -->

mohsin-r reviewed (approved):
> Reviewed 7 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2314)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @james-rae)



<!-- Sent from Reviewable.io -->
